### PR TITLE
[Microsoft SQL server] Added the Errorlog datastream for Microsoft SQL server

### LINKED ISSUE #3547: MS SQL Logs datastream migration to Agent
### Enhancement:

- The goal of this initiative is to migrate the existing observability support for MS SQL server error logs.
- As an Elastic user, I want the ability to collect error logs from my MS SQL server nodes and store the data in Elasticsearch so that I could visualize and analyze those logs easily whenever needed.
- As an Elastic integrations user, I would like to provide the following config options so that I can control how the Elastic agent connects and collects logs from my MS SQL server nodes.
Paths: A set of absolute paths that specify where to look for the log files.

### User Experience:

- Go to the ‘Homepage’ of the Elastic Cloud.
- Click on the ‘Add Integrations’ button available on the home page or on the side menu to navigate to the ‘Integrations’ page.
- Scroll through the list or search for a related keyword or click on the datastore filter to find the tile with the title ‘Microsoft SQL Server’ and description ‘Collect logs and metrics from Microsoft SQL Server instances with Elastic Agent.’
- Click on the given tile and navigate to the Overview page of the Microsoft SQL Server integration.
- On the ‘Overview’ page, click on the ‘Add Microsoft SQL Server’ button, to see the configuration parameters and build an integration policy.
- On the integration policy page, enable/disable the toggle titled ‘Collect error logs from Microsoft SQL server’.
- Scroll to the bottom of the page and click on the ‘Save and Continue’ button.

### All changes

- [x] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md)
- [x] Supported versions of the monitoring target are documented
- [x] Supported operating systems are documented (if applicable)
- [x] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist
- [x] [Pipeline](https://github.com/elastic/elastic-package/blob/main/docs/howto/pipeline_testing.md) test exist.
- [x] Documentation exists
- [x] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
- [x] At least a manual test with ES / Kibana / Agent has been performed.
- [x] Required Kibana version set to:

### PR DESCRIPTION
**Pull request by ManojS-shetty**
- Enhancement

## What does this PR do?

This PR adds the log datastream for Microsoft_sqlserver integration package. The Microsoft SQL Server `errorlog` contains user-defined events and certain system events you can use for troubleshooting.

See: [View the SQL Server error log in SQL Server Management Studio](https://docs.microsoft.com/en-us/sql/relational-databases/performance/view-the-sql-server-error-log-sql-server-management-studio?view=sql-server-ver16)

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

- Clone the PR branch to local development tool.
- Build the ``/package/microsoft_sqlserver/`` using ``elastic-package`` command-line tool.
- Enable the Collect logs from Microsoft SQL server and provide the Errorlog file path generated by Microsoft SQL server instance.
- Save the integration, once saved click on the assets to view the visualisation.
- Run the pipeline test and system test using ``elastic-package`` tool.

## Related issues

- Closes #3547

## Screenshots
Screenshot 2022-07-08 at 11 28 14 AM

### CODE REVIEW

File: packages/microsoft_sqlserver/_dev/deploy/docker/docker-compose.yml [hunk 1]
```diff
@@ -2,9 +2,12 @@ version: '2.3'
 services:
   microsoft_sqlserver:
     image: mcr.microsoft.com/mssql/server:2019-latest
+    user: root
```
Comment (efd6) [inline packages/microsoft_sqlserver/_dev/deploy/docker/docker-compose.yml:5]: Why is this needed?

File: packages/microsoft_sqlserver/data_stream/log/_dev/test/pipeline/test.log-expected.json [hunk 2]
```diff
@@ -0,0 +1,2991 @@
...
+            },
+            "message": "Microsoft SQL Server 2019 (RTM-CU16-GDR) (KB5014353) - 15.0.4236.7 (X64) \n\tMay 29 2022 15:55:47 \n\tCopyright (C) 2019 Microsoft Corporation\n\tDeveloper Edition (64-bit) on Linux (Ubuntu 20.04.4 LTS) \u003cX64\u003e",
+            "microsoft_sqlserver": {
+                "log": {
+                    "origin": "Server"
+                }
+            },
+            "mssql": {
+                "log": {}
+            },
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/_dev/test/pipeline/test.log-expected.json:27]: Remove these empty objects?

File: packages/microsoft_sqlserver/data_stream/log/agent/stream/log.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  - {{tag}}
+{{/each}}
+{{#contains "forwarded" tags}}
+publisher_pipeline.disable_host: true
+{{/contains}}
+exclude_files: [".gz$"]
+multiline:
+  pattern: '^\d{4}-\d{2}-\d{2}'
+  negate: true
+  match: after
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/agent/stream/log.yml.hbs:19]: 😄

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+- grok:
+    field: message
+    patterns:
+    - '%{MSSQL_DATE:date} %{DATA:mssql.log.origin} [ ]*%{GREEDYDATA:msg_temp}'
+    - '%{GREEDYDATA:msg_temp}'
+    pattern_definitions:
+      MSSQL_DATE: '%{DATA} %{DATA}'
+      GREEDYDATA: |-
+          (.|
+          |	)*
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml:19]: This looks exceedingly brittle to accidental editor damage to me. Can you use escape sequences, please.
```suggestion
      GREEDYDATA: '(.|\n)*'
```
Note that the `\t` is already matched by the `.`, only the newline is special (this change is confirmed locally).

File: packages/microsoft_sqlserver/_dev/deploy/docker/docker-compose.yml (hunk 1 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/_dev/deploy/docker/docker-compose.yml:5]: As there was access denied error message in container logs only particular in CI, need to get this docker compose file changed a little and CI got green.

File: packages/microsoft_sqlserver/data_stream/log/_dev/test/pipeline/test.log-expected.json (hunk 2 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/_dev/test/pipeline/test.log-expected.json]: Done.

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    on_failure:
+    - append:
+        field: error.message
+        value: '{{ _ingest.on_failure_message }}'
+- remove:
+    field: date
+    ignore_missing: true
+- rename:
+    field: message
+    target_field: event.original
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml:43]: I'd move this processor up to the top with an `ignore_missing: true` and change the uses of `message` to `event.original`. This also means you can grok the greedy data directly into `message` and remove the rename `msg_temp` => `message`.

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- rename:
+    field: mssql.log.origin
+    target_field: microsoft_sqlserver.log.origin
+    ignore_missing: true
+    ignore_failure: true
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml:48]:
```suggestion
- rename:
    field: mssql
    target_field: microsoft_sqlserver
    ignore_missing: true
```

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+- append:
+    field: event.category
+    value: database
+- append:
+    field: event.type
+    value: info
+- remove:
+    field: mssql
+    ignore_missing: true
+    ignore_failure: true
```
Comment (efd6) [inline packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml:65]: Delete this since the rename has happened.

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 5 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: Done

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 6 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: Done

File: packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: Done

**Review by efd6 — APPROVED** [review]

File: packages/microsoft_sqlserver/data_stream/log/fields/agent.yml [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      description: 'Name of the host.
+
+        It can contain what `hostname` returns on Unix systems, the fully qualified domain name, or a name specified by the user. The sender decides which value to use.'
+    - name: os.family
+      level: extended
+      type: keyword
+      ignore_above: 1024
+      description: OS family (such as redhat, debian, freebsd, windows).
+      example: debian
+    - name: os.kernel
```
[the remark this answers is not in the document]
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/fields/agent.yml:146]: This change needs to be taken with wider integrations as the agent.yml is available across all integration.

File: packages/microsoft_sqlserver/data_stream/log/manifest.yml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        type: bool
+        multi: false
+        default: false
+      - name: processors
+        type: yaml
+        title: Processors
+        multi: false
+        required: false
+        show_user: false
+        description: >
```
Comment (muthu-mps) [inline packages/microsoft_sqlserver/data_stream/log/manifest.yml:38]: Can we have the description in the same line as we have in line-28 ?

File: packages/microsoft_sqlserver/data_stream/log/fields/base-fields.yml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+- name: event.dataset
+  type: constant_keyword
+  description: Event dataset
+  value: microsoft_sqlserver.log
+- name: input.type
+  description: Type of Filebeat input.
+  type: keyword
+- name: log.flags
+  type: keyword
+  description: Log flags.
```
Comment (muthu-mps) [inline packages/microsoft_sqlserver/data_stream/log/fields/base-fields.yml:23]: Please add a single line description about the flag thats captured.

File: packages/microsoft_sqlserver/data_stream/log/manifest.yml (hunk 9 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/manifest.yml]: Done

File: packages/microsoft_sqlserver/data_stream/log/fields/base-fields.yml (hunk 10 above)
Comment (ManojS-shetty) [reply packages/microsoft_sqlserver/data_stream/log/fields/base-fields.yml]: Done

**Review by muthu-mps — COMMENTED** [review]
Can we rename the Dashboard and visualisation title ?
_[Logs Microsoft SQL Server Error Log]_ Overview can be renamed as `[Logs Microsoft SQL Server] Error Log Overview`

**Comment by ManojS-shetty** [comment]
> Can we rename the Dashboard and visualisation title ? _[Logs Microsoft SQL Server Error Log]_ Overview can be renamed as `[Logs Microsoft SQL Server] Error Log Overview`

Done.

**Review by muthu-mps — APPROVED** [review]
LGTM!
